contracts/zero-ex/contracts/src/features/multiplex/MultiplexFeature.sol

```diff
@@ -0,0 +1,725 @@
+                (address[] memory tokens, bool isSushi) = abi.decode(
+                    subcall.data,
+                    (address[], bool)
+                );
+                target = _computeUniswapPairAddress(
+                    tokens[0],
+                    tokens[1],
+                    isSushi
+                );
+            } else if (subcall.id == MultiplexSubcall.LiquidityProvider) {
+                // Similar to UniswapV2, LiquidityProvider contracts
+                // allow tokens to be transferred in before the swap
+                // is executed, so we the target is the address encoded
+                // in the subcall data.
+                (target,) = abi.decode(
+                    subcall.data,
+                    (address, bytes)
+                );
+            } else if (
+                subcall.id == MultiplexSubcall.UniswapV3 ||
+                subcall.id == MultiplexSubcall.BatchSell
+            ) {
+                // UniswapV3 uses a callback to pull in the tokens being
+                // sold to it. The callback implemented in `UniswapV3Feature`
+                // can either:
+                // - call `transferFrom` to move tokens from `msg.sender` to the
+                //   UniswapV3 pool, or
+                // - call `transfer` to move tokens from `address(this)` to the
+                //   UniswapV3 pool.
+                // A nested batch sell is similar, in that it can either:
+                // - use tokens from `msg.sender`, or
+                // - use tokens held by `address(this)`.
+
+                // Suppose UniswapV3/BatchSell is the first call in the multi-hop
+                // path. The input tokens are either held by `msg.sender`,
+                // or in the case of `multiplexMultiHopSellEthForToken` WETH is
+                // held by `address(this)`. The target is set accordingly.
+
+                // If this is _not_ the first call in the multi-hop path, we
+                // are dealing with an "intermediate" token in the multi-hop path,
+                // which `msg.sender` may not have an allowance set for. Thus
+                // target must be set to `address(this)` for `i > 0`.
+                if (i == 0 && !params.useSelfBalance) {
+                    target = msg.sender;
+                } else {
+                    target = address(this);
+                }
+            } else {
+                revert("MultiplexFeature::_computeHopTarget/INVALID_SUBCALL");
+            }
+        }
+        require(
+            target != address(0),
+            "MultiplexFeature::_computeHopTarget/TARGET_IS_NULL"
+        );
+    }
+
+    // If `rawAmount` encodes a proportion of `totalSellAmount`, this function
+    // converts it to an absolute quantity. Caps the normalized amount to
+    // the remaining sell amount (`totalSellAmount - soldAmount`).
+    function _normalizeSellAmount(
+        uint256 rawAmount,
+        uint256 totalSellAmount,
```

Comment:
oh I see hmmm
I wonder how often we'd be able to recover from one leg failing though, since the remaining sources would still need to satisfy `minBuyAmount`. Maybe if we were to order the fills in increasing order of proportion? We could achieve something pretty close to this with proportional fills by leaving the largest-proportion fill at the end and giving at a proportion of 100%.

Reply:
Yeah that's what I was thinking, but it's not a priority since we aren't using proportional fills anywhere.

Reply:
going to keep as is for now then